### DISCUSSION TITLE
I2s rx how to determine proper audio

### DISCUSSION THREAD

**Discussion by aqildad-create** [pschatzmann/arduino-audio-tools#826] [Q&A] [answered]
How to determine if i  have  a signal on i2s so to send data to  ESPNow ACCORDNGLY . i want to avoid sending empty data 

You do not have example for VolumeOutput  . Pls guide

**Comment by pschatzmann** [comment, marked as answer]
You can use the VolumeOutput to determine if you have a signal and based on this, switch the forwarding to the ESPNow on or off.

```
                                      -> VolumeOutput   
I2S - copy -> MultiOutput   
                                      -> OnOffOutput -> ESPNowStream
```
VolumeOutput is just like any output destination, so no surprise here. 
Check the class documentation for the methods !

**Reply by bfulham** [reply]
does the OnOffOutput still exist? I cant seem to find any reference to it

**Reply by bfulham** [reply]
did some more digging and found you commented it out in feb then removed it all together in march? did it get replaced with an alternative way of doing this?

**Comment by pschatzmann** [comment]
Not sure why it has been removed: Most likely because the same thing can be achieved with a VolumeStream: setting the volume to 0.0 stops the output, setting it to 1.0 activates it again (to it's maximum)

**Reply by bfulham** [reply]
Ive been trying to make a device that can switch from an input (input 1) to a second input (input 2) when input 2 detects audio, ive got it 95% working but I am struggling joining the 2 streams back up again after, Im currently doing this

BluetoothA2DPSink -> VolumeStream -> AudioBoardStream

as well as
                                                      
AudioBoardStream -> StreamCopy -> MultiOutput 
MultiOutput -> VolumeOutput
MultiOutput -> VolumeStream -> AudioBoardStream

It works perfectly but as I expected I am getting a heap of stuttering, is there a proper way to turn streams on and off or swap between inputs to the AudioBoardStream?

**Reply by bfulham** [reply]
```
#include "AudioTools.h"
#include "AudioLibs/A2DPStream.h"
#include "AudioLibs/AudioBoardStream.h"

AudioInfo info(44100, 2, 16);
BluetoothA2DPSink ble;
AudioBoardStream kit(AudioKitEs8388V1);
VolumeStream UHFOnOff(kit);
VolumeStream BLEOnOff(kit);
VolumeOutput vMeter;
MultiOutput uhfmulti(vMeter, UHFOnOff);
StreamCopy copy_UHF(uhfmulti, kit); // copy kit to uhfmulti

bool uhf = false;

// Write data to AudioKit in callback
void read_data_stream(const uint8_t *data, uint32_t length) {
    BLEOnOff.write(data, length);
}

// Setup
void setup(void) {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // setup the audioKit
  auto cfg = kit.defaultConfig(RXTX_MODE);
  cfg.sd_active = false;
  cfg.input_device = ADC_INPUT_LINE2;
  cfg.copyFrom(info);
  
  // register callback
  ble.set_stream_reader(read_data_stream, false);
  ble.set_auto_reconnect(true, 1000);

  kit.begin(cfg);
  vMeter.begin(info);
  BLEOnOff.begin();
  UHFOnOff.begin();
  ble.start("AudioKit");
}

// loop  
void loop() {
  copy_UHF.copy();
	if(vMeter.volume() > 200 && uhf == false){
    uhf = true;
    BLEOnOff.setVolume(0);
    UHFOnOff.setVolume(1);
    Serial.println("uhf");
  } else if (vMeter.volume() < 200 && uhf == true) {
    uhf = false;
    UHFOnOff.setVolume(0);
    BLEOnOff.setVolume(1);
    Serial.println("ble");
  }
}
```

**Reply by bfulham** [reply]
@pschatzmann?

**Reply by pschatzmann** [reply]
I think in your case, it would be better to work just with your bool variable uhf to switch things on and off.
So remove VolumeStream UHFOnOff(kit) and VolumeStream BLEOnOff(kit);

And in your callback use if (uhf) ble.write(data, length); and in your loop use  if (!uhf)  copy_UHF.copy();

**Reply by bfulham** [reply]
I hadn’t thought about using that callback, I’ll try that, I originally was using two copy’s and changed which one was copying but that had performance issues

**Reply by bfulham** [reply]
is there another way to get the ausio to the multiOutput that isnt using the streamcopy? implementing this I still have to have a streamcopy in the loop that isnt covered by the if (uhf) as it needs to get the audio to the volumeOutput to know when to change the uhf variable, meaning the uhf audio work perfectly but the bluetooth audio still stutters because of the copy, it has to look like this:

AudioBoardStream -> StreamCopy1 -> MultiOutput
MultiOutput -> VolumeOutput
MultiOutput -> streamcopy2 -> AudioBoardStream

streamcopy1 is always running while 2 gets turned on and off with uhf

**Reply by pschatzmann** [reply]
You can simple reconfigure the copy to switch between MultiOutput and VolumeOutput as copy destination.

**Reply by bfulham** [reply]
you are a life saver, this works perfectly, once i finish the project im going to write up a show and tell for it
